Add type annotations and refactor to use TokenDict type

rdimaio opened the pull request:
#6454 - Adding various type annotations, and replacing the method:

```
def token_dictionary(token: models.Token):
    return {'token': token.token, 'expires_at': token.expired_at}
```

with a specific datatype `TokenDict` in order to be more specific with the return types.

dchristidis commented on lib/rucio/api/authentication.py:
```diff
@@ -230,14 +230,13 @@ def get_auth_token_ssh(account, signature, appid, ip=None, vo='def', *, session:
 
 @transactional_session
-def get_ssh_challenge_token(account, appid, ip=None, vo='def', *, session: "Session"):
+def get_ssh_challenge_token(account, ip=None, vo='def', *, session: "Session"):
```
This is quite a significant change that is currently invisible (think of a community that has implemented probes using the API layer, as is the official recommendation). Anything related to the removal of `appid` must be split into its own commit.

dchristidis replied:
Please review this again. It looks to me like you reverted the change in `core/authentication.py`, but not elsewhere.

(On a side note, I know you’re trying to be helpful, but I find it more beneficial to resolve conversations myself. This way, when I resolve it, I know for sure that it’s done and have no need to check it again. No need to reply or anything to demonstrate that you’ve done your part. Thanks!)

dchristidis replied:
Still seeing changes related to this in `lib/rucio/web/rest/flaskapi/v1/auth.py` and `
tests/test_authentication.py`.

dchristidis commented on lib/rucio/common/types.py:
```diff
@@ -14,6 +14,7 @@
 # limitations under the License.
 
 from typing import Any, Callable, Optional, TypedDict, Union
+from datetime import datetime
```
Pleas sort alphabetically.

dchristidis commented on lib/rucio/common/utils.py:
```diff
@@ -570,7 +570,7 @@ def str_to_date(string):
 
 
-def val_to_space_sep_str(vallist):
+def val_to_space_sep_str(vallist: list) -> str:
```
Doesn’t the `vallist: list` hint contradict the implementation (which also does something if it isn’t a list)?

rdimaio replied:
Seems like `list` is the intended type for `vallist`, but instead of failling when a list is not passed, the implementation simply prints a string representation of whatever the object is. 

I think it would be better to instead raise an exception, as it would make it more explicit that `val_to_space_sep_str` is not actually returning a space-separated string. 

I don't want to make a functional change here, so I will just modify vallist to be `Any` for now, but we might want to consider modifying this implementation at some point.

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -17,7 +17,8 @@
 from traceback import format_exc
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Any, Iterator, Optional
+import uuid
```
Importing entire modules should be placed above selective imports with `from`.

dchristidis replied:
You placed `uuid` correctly, but also introduced an unnecessary empty newline.

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -37,7 +39,7 @@
 
 @transactional_session
-def add_account(account, type_, email, *, session: "Session"):
+def add_account(account: InternalAccount, type_: AccountType, email: str, *, session: "Session"):
```
Missing hint for the return type.

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -102,7 +104,7 @@ def get_account(account, *, session: "Session"):
 
 @transactional_session
-def del_account(account, *, session: "Session"):
+def del_account(account: InternalAccount, *, session: "Session"):
```
Missing hint for the return type.

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -115,15 +117,14 @@ def del_account(account, *, session: "Session"):
         raise exception.AccountNotFound('Account with ID \'%s\' cannot be found' % account)
 
-    account.update({'status': AccountStatus.DELETED, 'deleted_at': datetime.utcnow()})
-
```
You’ve also moved around statements later as well so that they are part of the try block. Please elaborate.

rdimaio replied:
The original statement `account = session.execute(query).scalar_one()` would lead to a type error, because the type of `account` (`InternalAccount`) was not compatible with the type returned by `session.execute(query).scalar_one()` (I don't remember what this type is at the moment). 

To avoid this, I changed the statemente so that the output of `session.execute(query).scalar_one()` would be stored in a new variable `account_result`. 

As this variable is defined within the scope of the `try` block, I had to move the `update` statement under the `try` block as well.

dchristidis replied:
I’m not sure I follow. The `try` block does not create a new scope.

```python
>>> print(foo)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'foo' is not defined
>>> try:
...     foo = 'foo'
... except Exception:
...     pass
... 
>>> print(foo)
foo
>>> 
```

rdimaio replied:
You're right - fixed

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -115,15 +117,14 @@ def del_account(account, *, session: "Session"):
 
 @transactional_session
-def update_account(account, key, value, *, session: "Session"):
+def update_account(account: InternalAccount, key: str, value: Any, *, session: "Session"):
```
Missing hint for the return type.

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -115,15 +117,14 @@ def del_account(account, *, session: "Session"):
 
 @transactional_session
-def update_account(account, key, value, *, session: "Session"):
+def update_account(account: InternalAccount, key: str, value: Any, *, session: "Session"):
```
Like my previous comment, the implementation suggests that value may not be a `str`.

rdimaio replied:
`value` is set to `Any` here, not `str`. Could you clarify what you mean?

dchristidis replied:
You’re right. I’m not sure what I was thinking; probably a mistake on my part.

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -137,22 +138,22 @@ def update_account(account, key, value, *, session: "Session"):
 
 @stream_session
-def list_accounts(filter_=None, *, session: "Session"):
+def list_accounts(filter_: Optional[dict] = None, *, session: "Session") -> Iterator[dict]:
```
Maybe `filter_: Optional[dict[str, Any]] = None` is more precise?

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -213,7 +214,7 @@ def list_accounts(filter_=None, *, session: "Session"):
 
 @read_session
-def list_identities(account, *, session: "Session"):
+def list_identities(account: InternalAccount, *, session: "Session"):
```
Missing hint for the return type.

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -248,7 +249,7 @@ def list_identities(account, *, session: "Session"):
 
 @read_session
-def list_account_attributes(account, *, session: "Session"):
+def list_account_attributes(account: InternalAccount, *, session: "Session") -> list[dict]:
```
Same comment: consider `-> list[dict[str, Any]]`

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -298,7 +299,7 @@ def has_account_attribute(account, key, *, session: "Session"):
 
 @transactional_session
-def add_account_attribute(account, key, value, *, session: "Session"):
+def add_account_attribute(account: InternalAccount, key: str, value: Any, *, session: "Session"):
```
Missing hint for the return type.

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -334,7 +335,7 @@ def add_account_attribute(account, key, value, *, session: "Session"):
 
 @transactional_session
-def del_account_attribute(account, key, *, session: "Session"):
+def del_account_attribute(account: InternalAccount, key: str, *, session: "Session"):
```
Missing hint for the return type.

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -355,7 +356,7 @@ def del_account_attribute(account, key, *, session: "Session"):
 
 @read_session
-def get_usage(rse_id, account, *, session: "Session"):
+def get_usage(rse_id: uuid.UUID, account: InternalAccount, *, session: "Session") -> dict:
```
I don’t think this is right. `rse_id` should be `str`. This applies to many other places as well.

rdimaio replied:
I was looking at this statement within the function:
```
models.AccountUsage.rse_id == rse_id
```

Where the model is:

```
class AccountUsage(BASE, ModelBase):
    """Represents account usage"""
    __tablename__ = 'account_usage'
    account: Mapped[InternalAccount] = mapped_column(InternalAccountString(get_schema_value('ACCOUNT_LENGTH')))
    rse_id: Mapped[uuid.UUID] = mapped_column(GUID())
...
```

So I assumed `rse_id` would need to match the type indicated in the model

dchristidis replied:
Good assumption, but it’s complicated. You can see that SQLAlchemy will always return `Optional[str]` when querying the database:

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/db/sqla/types.py#L70-L78

So in the API layer where the RSE name is converted to an RSE ID:

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/api/rse.py#L333-L334

It is most definitely a `str`.

rdimaio replied:
I changed all of them to `str`, it only caused this error:

```
Found 2 new problems in /home/runner/work/rucio/rucio/lib/rucio/core/account_limit.py
  - 1 errors with message """Expression of type "Unknown | UUID" cannot be assigned to declared type "str | None"""".
    Candidates: line 293
  - 1 errors with message """Argument of type "str | None" cannot be assigned to parameter "__key" of type "UUID" in function "get"
                               Type "str | None" cannot be assigned to type "UUID"
                                 "str" is incompatible with "UUID"""".
    Candidates: line 294
```

under `get_local_account_usage`. I've now annotated that `rse_id` to be `uuid.UUID`, and left all the others to be `str`.

dchristidis replied:
I have the impression that it would be more correct to change the model to have `rse_id: Mapped[str]`. The `GUID` class literally has `impl = CHAR`. I’m consulting the SQLAlchemy documentation, in hope of finding a definitive answer. Could you also investigate this path?

rdimaio replied:
> I have the impression that it would be more correct to change the model to have `rse_id: Mapped[str]`. The `GUID` class literally has `impl = CHAR`. I’m consulting the SQLAlchemy documentation, in hope of finding a definitive answer. Could you also investigate this path?

I'm not sure about making this change - looking at the other models, all `id` fields seem to use `uuid.UUID`. For consistency, I think it would be best to keep this one as `uuid.UUID` as well.

dchristidis replied:
If `Mapped[str]` is correct for `mapped_column(GUID())`, then, by definition, all instances of `Mapped[uuid.UUID]` are incorrect and will have to be replaced.

rdimaio replied:
Addressed in https://github.com/rucio/rucio/pull/6497/commits/c20301e65f7f1a8c0c886beebdb4f17d01958d46

rdimaio replied:
Update - this causes `test_lifetime_creation_core` to fail:

```
E   psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "CONFIGS_PK"
E   DETAIL:  Key (section, opt)=(lifetime_model, cutoff_date) already exists.
```

rdimaio replied:
seems like it was a one-off issue - it passed when rerunning the test https://github.com/rucio/rucio/actions/runs/8005839275/job/21866803841?pr=6497#step:4:917

dchristidis commented on lib/rucio/core/account_counter.py:
```diff
@@ -12,23 +12,26 @@
 # limitations under the License.
+
+import uuid
 import datetime
```
Please sort alphabetically.

dchristidis commented on lib/rucio/core/account_counter.py:
```diff
@@ -82,7 +85,7 @@ def del_counter(rse_id, account, *, session: "Session"):
 
 @read_session
-def get_updated_account_counters(total_workers, worker_number, *, session: "Session"):
+def get_updated_account_counters(total_workers: int, worker_number: int, *, session: "Session") -> list["Query"]:
```
`-> list["Query"]` doesn’t look right.

rdimaio replied:
This was the reference I used for that: https://stackoverflow.com/questions/55361186/type-hinting-sqlalchemy-query-result

> [Query.all()](https://github.com/sqlalchemy/sqlalchemy/blob/201c4a60e4b8af56d9c02a3675d1443ba4171c89/lib/sqlalchemy/orm/query.py#L3155-L3161) calls list() on the Query object itself

I'm not 100% sure about it either, but I'm not sure about what other type this would be.

dchristidis replied:
But looking at the query:

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/core/account_counter.py#L94

I would have guessed something like `list[tuple[InternalAccount, str]]`. Did you try it and saw errors?

rdimaio replied:
I have not tried `list[tuple[InternalAccount, str]]` yet - I'll try it now to see if it works

dchristidis replied:
We need to have a closer look at this. `.all()` returns instances of `sqlalchemy.engine.row.Row`. I’m thinking it would be preferable to:

 1. Convert the query to SQLAlchemy 2.0 syntax (unrelated to hints, but a good opportunity)
 2. Return dictionaries with `return [row._asdict() for row in query.all()]`
 3. Adapt `lib/rucio/daemons/abacus/account.py`

My motivation for this proposal is that the use of indices feels poor:

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/daemons/abacus/account.py#L72-L73

What do you think?

And on a side note, do you understand the usefulness of this? I think it’s identical and can go in `filter_thread_work()` directly.

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/core/account_counter.py#L97-L100

It was introduced in f3a073aa9da181070d6b41572838d28727528dd6.

rdimaio replied:
I've done the first three changes suggested in the latest revision, but I think they're outside the scope of this PR, and I think the focus should be kept on typing (and make separate PRs for these behavioural changes when needed, which I think might happen fairly frequently with these typing-focused PRs) 

On `filter_thread_work`: I agree that it could be refactored like that, but it seems like that method is used in quite a few other places and I don't think refactoring would fit the scope of this PR - best to open an issue for it and we can work on it separately.

dchristidis replied:
Yes and no.

The SQLAlchemy 2.0 syntax, I admit, was over the top. There are five trivial queries in that file; they can all be worked on at the same time in a PR which aims to do it from the start. In retrospect, I should not have asked you to do it. I’ll let you choose whether to keep it or not. Note, however, that you’ll have to follow the coding style we’ve come to adopt for queries:

```python
query = select(
    models.UpdatedAccountCounter.account,
    models.UpdatedAccountCounter.rse_id,
).distinct(
)
```

Now, the change in the return type of `get_updated_account_counters()` *is* very much related to this PR. I’ve asked you to do it because I believe `list[tuple[InternalAccount, str]]` is incorrect. So we had three options:

 1. Not add type hints.
 2. Add incorrect type hints.
 3. Adjust and add correct type hints.

I chose option (3). However, you cannot hide it under `Auth: Migrate to SQLAlchemy 2.0 syntax; #6057`. You have to split it into its own commit.

Now, I did not understand your comment about `filter_thread_work()`. That function is completely fine as it is. What I am asking you to do is to:

 1. Remove those four lines.
 2. Replace `hash_variable=hash_variable` with `hash_variable='CONCAT(account, rse_id)'`.

~~In a separate commit, of course. Yes, it’s not related to type hints, but unlike the SQLAlchemy 2.0 syntax migration, there will be no dedicated issue or pull request to fix it (because it doesn’t happen anywhere else).~~

PS: Actually, we just have to not forget to apply this correction together with the 2.0 syntax change. So it ties together with the choice I mentioned at the beginning of this comment. I would have asked you to remove it, except for the fact that you’ve already done most of it.

rdimaio replied:
I had misunderstood regarding `filter_thread_work()`, I thought you meant that this logic:

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/core/account_counter.py#L97-L100

could have gone directly within the `filter_thread_work()` function, rather than being done outside of it; now I understood that you meant that the `CONCAT` command is case insensitive and thus the `if/else` block is not needed. I'll make the change you proposed now

rdimaio replied:
Regarding the query syntax - I opened this documentation issue https://github.com/rucio/documentation/issues/287 to consider adding guidelines like these to the documentation

dchristidis commented on lib/rucio/core/authentication.py:
```diff
@@ -530,12 +530,8 @@ def validate_auth_token(token: str, *, session: "Session") -> "dict[str, Any]":
-def token_dictionary(token: models.Token):
-    return {'token': token.token, 'expires_at': token.expired_at}
-
-
```
Curious, why is removing `token_dictionary()` preferable? I’m not necessarily against it, but seems like a useful function.

rdimaio replied:
I can keep both the function and `TokenDict` (and just make `token_dictionary` return a `TokenDict` dict instead of a generic dict) but it seemed a bit unnecessary to make a function just for that. I can add it again if you prefer.

dchristidis replied:
It’s true that with hints you can be confident that the structure of the returned value will be correct, *if you’re testing for type checking*. But what about if someone accidentally does something like:

```python
return {'token': token.token, 'expires_at': datetime.utcnow()}
```

Far-fetched, I know, but hopefully it demonstrates that this function still has some value and is effectively more strict than just type checking.

dchristidis replied:
Are you purposefully changing the location of `token_dictionary()` in the file?

rdimaio replied:
Fixed in latest revision

dchristidis replied:
And what about `lib/rucio/core/oidc.py`?

dchristidis commented on lib/rucio/core/authentication.py:
```diff
@@ -530,12 +530,8 @@ def validate_auth_token(token: str, *, session: "Session") -> "dict[str, Any]":
-
 @transactional_session
-def __delete_expired_tokens_account(account, *, session: "Session"):
+def __delete_expired_tokens_account(account: InternalAccount, *, session: "Session"):
```
Missing hint for the return type.

dchristidis commented on lib/rucio/core/oidc.py:
```diff
@@ -1000,7 +1000,7 @@ def __change_refresh_state(token: str, refresh: bool = False, *, session: "Sessi
 
 @transactional_session
-def refresh_cli_auth_token(token_string: str, account: str, *, session: "Session"):
+def refresh_cli_auth_token(token_string: str, account: str, *, session: "Session") -> Optional[tuple]:
```
Why not be more precise with `-> Optional[tuple[str, int]]`?

dchristidis reviewed (changes requested):
Looks promising! It would be beneficial to prefix the commit subjects with `Auth: `.

rdimaio commented on lib/rucio/core/account_counter.py:
```diff
@@ -91,21 +93,24 @@ def get_updated_account_counters(total_workers, worker_number, *, session: "Sess
         hash_variable = 'concat(account, rse_id)'
 
     query = filter_thread_work(session=session, query=query, total_threads=total_workers, thread_id=worker_number, hash_variable=hash_variable)
+    query_result = session.execute(query)
```
not 100% sure here if I had to filter `query` or `query_result`

dchristidis replied:
`filter_thread_work()` takes a query and returns another one. Think of it as building on the existing query, like:

```python
query = query.where(models.UpdatedAccountCounter.rse_id == …)
```

I would like to question the introduction of `query_result` (which can be somewhat misleading because it does not contain any result. Who you oppose the following?

```python
return [row._asdict() for row in session.execute(query).all()]
```

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -137,22 +138,22 @@ def update_account(account, key, value, *, session: "Session"):
-        elif value == AccountStatus.ACTIVE:
-            account.update({'status': value, 'suspended_at': None})
-    else:
-        account.update({key: value})
```
Similar to another comment, I do not understand the need for this code rearrangement.

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -137,22 +138,22 @@ def update_account(account, key, value, *, session: "Session"):
 
 @stream_session
-def list_accounts(filter_=None, *, session: "Session"):
+def list_accounts(filter_: Optional[dict[str, Any]] = None, *, session: "Session") -> Iterator[dict]:
```
Do we want to prefer `Iterator[dict[str, Any]]`? We should aim to eventually replace all `dicts` with `TypedDict`s or the model themselves, but until then, is being a tiny bit more specific beneficial? This comment applies to other lines (and other files) too.

dchristidis commented on lib/rucio/core/account_counter.py:
```diff
@@ -28,7 +30,7 @@
 
 @transactional_session
-def add_counter(rse_id, account, *, session: "Session"):
+def add_counter(rse_id: str, account: InternalAccount, *, session: "Session"):
```
Missing `-> None`. Same below. Same for the very last function of this file.

dchristidis commented on lib/rucio/core/account_limit.py:
```diff
@@ -196,7 +197,7 @@ def get_local_account_limits(account, rse_ids=None, *, session: "Session"):
 
 @transactional_session
-def set_local_account_limit(account, rse_id, bytes_, *, session: "Session"):
+def set_local_account_limit(account: InternalAccount, rse_id: str, bytes_: int, *, session: "Session"):
```
Missing `-> None`. Also below.

dchristidis commented on lib/rucio/core/authentication.py:
```diff
@@ -40,7 +41,7 @@
 if TYPE_CHECKING:
     from sqlalchemy.orm import Session
-    from typing import Any, Union
+    from typing import Any
```
We no longer recommending `Any` being imported conditionally. I would recommend moving it up with the others, then changing all uses of it (the hints to not be quoted in strings).

dchristidis commented on lib/rucio/core/authentication.py:
```diff
@@ -214,7 +215,7 @@ def get_auth_token_gss(account, gsstoken, appid, ip=None, *, session: "Session")
 
 @transactional_session
-def get_auth_token_ssh(account, signature, appid, ip=None, *, session: "Session"):
+def get_auth_token_ssh(account: InternalAccount, signature: Union[str, bytes], appid: str, ip: Optional[str] = None, *, session: "Session") -> Optional[TokenDict]:
```
I think `Union[str, bytes]` may warrant one additional look. As far as I can tell, the current calls of this functions use `bytes`. So why keep supporting `str`?

dchristidis commented on lib/rucio/core/oidc.py:
```diff
@@ -1444,5 +1444,5 @@ def oidc_identity_string(sub: str, iss: str):
 
 
-def token_dictionary(token: models.Token):
+def token_dictionary(token: models.Token) -> TokenDict:
```
Is there something preventing us from reusing the function in `authentication.py`?

dchristidis reviewed (changes requested):
I will also like to see some rework on how the changes are split into commits, but it might be easier to talk about it in person.

rdimaio commented on lib/rucio/core/oidc.py:
```diff
@@ -1405,7 +1404,7 @@ def validate_jwt(json_web_token: str, *, session: "Session") -> dict[str, Any]:
                                           % (oidc_client.client_id, oidc_client.client_secret),
                                           '-d', 'token=%s' % (json_web_token),
-                                          oidc_client.introspection_endpoint],
+                                          oidc_client.introspection_endpoint],  # type: ignore
```
note: this one is pending discussion on Mattermost to decide whether we can delete this whole if block or not

rdimaio replied:
discussed - best to ignore / avoid changing anything OIDC-related

dchristidis commented on lib/rucio/common/types.py:
```diff
@@ -13,8 +13,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from uuid import UUID
```
Why are you reintroducing `UUID`? I thought we agreed that the correct type is `str`?

rdimaio replied:
Removed it from the typeddict. Will only keep it where the tests fails, i.e.:

```
Found 2 new problems in /home/runner/work/rucio/rucio/lib/rucio/core/account_limit.py
  - 1 errors with message """Expression of type "Unknown | UUID" cannot be assigned to declared type "str | None"""".
    Candidates: line 293
  - 1 errors with message """Argument of type "str | None" cannot be assigned to parameter "__key" of type "UUID" in function "get"
                               Type "str | None" cannot be assigned to type "UUID"
                                 "str" is incompatible with "UUID"""".
    Candidates: line 294
```

dchristidis commented on lib/rucio/common/utils.py:
```diff
@@ -1200,7 +1200,7 @@ def detect_client_location():
 
 
-def ssh_sign(private_key, message):
+def ssh_sign(private_key: str, message: Union[str, bytes]) -> str:
```
Do you see any current uses of this function where the `message` argument is `bytes`?

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -27,6 +27,7 @@
 import rucio.core.rse
 from rucio.common import exception
 from rucio.common.config import config_get_bool
+from rucio.common.types import InternalAccount, AccountAttributesDict, AccountDict, AccountUsageModelDict, IdentityDict, UsageDict
```
Please use `isort` if you don’t want to do the work manually. :smile:

dchristidis commented on tests/test_authentication.py:
```diff
@@ -146,7 +146,7 @@ def test_get_auth_token_ssh_fail(self, vo):
             pass  # might already exist, can skip
 
-        signature = ssh_sign(PRIVATE_KEY, 'sign_something_else')
+        signature = base64.b64decode(ssh_sign(PRIVATE_KEY, 'sign_something_else'))
 
         result = get_auth_token_ssh(account='root', signature=signature, appid='test', ip='127.0.0.1', vo=vo)
```
I’m not too fond of the situation we ended up in. `ssh_sign()` converts `bytes` to `str`, which you have to convert back to `bytes` to use `get_auth_token_ssh()`.

https://github.com/rucio/rucio/blob/e85183fe6b77841ca01e2d68a1426526bef81b1b/lib/rucio/web/rest/flaskapi/v1/auth.py#L1201-L1229

I wonder if it’s worth propagating the the payload as it is, as an `str`, down to the core layer. There, inside `get_auth_token_ssh()`, we put the decoding operation and raise let it raise a `TypeError` if it fails. The code in the REST layer can then be simplified.

But probably this is going too far, so feel free to ignore.

rdimaio replied:
Addressed in this follow-up PR: https://github.com/rucio/rucio/pull/6649

rdimaio committed:
Typing: Add types; #6454

rdimaio committed:
Auth: ssh_sign, only allow message as str

rdimaio committed:
Auth: get_auth_token_ssh, only allow signature in bytes

rdimaio committed:
Auth: Migrate to SQLAlchemy 2.0 syntax; #6057

rdimaio committed:
Typing: Add type hints; #6454

rdimaio commented:
@dchristidis I've addressed all your most recent comments and rebased the PR into separate commits in a way that makes sense. 

I've also had to rebase some stuff from master, so "Compare" shows some of those changes as well, which might make it hard to review the diff.

dchristidis commented on lib/rucio/common/utils.py:
```diff
@@ -40,7 +40,7 @@
 from io import StringIO
 from itertools import zip_longest
-from typing import TYPE_CHECKING, Callable, Optional
+from typing import Any, TYPE_CHECKING, Callable, Optional
```
`isort` is your friend!

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -244,11 +245,11 @@ def list_identities(account, *, session: "Session"):
     ).where(
         models.IdentityAccountAssociation.account == account
     )
-    return [row._asdict() for row in session.execute(query)]
+    return [cast(IdentityDict, row._asdict()) for row in session.execute(query)]
```
Not so much a comment for you, but more so for me: understand the necessity to use `cast()`.

rdimaio replied:
This is what occurs when I don't use `cast`:

```
Found 5 new problems in /home/runner/work/rucio/rucio/lib/rucio/core/account.py
  - 1 errors with message """Expression of type "list[Dict[str, Any]]" cannot be assigned to return type "list[IdentityDict]"""".
    Candidates: line 248
  - 1 errors with message """Expression of type "list[Dict[str, Any]]" cannot be assigned to return type "list[AccountAttributesDict]"""".
    Candidates: line 278
  - 1 errors with message """Expression of type "Dict[str, Any]" cannot be assigned to return type "UsageDict"
                               "Dict[str, Any]" is incompatible with "UsageDict"""".
    Candidates: line 377
  - 1 errors with message """Expression of type "list[dict[str, Any]]" cannot be assigned to return type "list[AccountUsageModelDict]"""".
    Candidates: line 398
  - 1 errors with message """Expression of type "list[Dict[str, Any]]" cannot be assigned to return type "list[UsageDict]"""".
    Candidates: line 424
```

See here for reference: https://github.com/rdimaio/rucio/actions/runs/8206991793/job/22447307300#step:9:10

dchristidis replied:
I think the `Dict[str, Any]` hint must be coming from [SQLAlchemy itself](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row._asdict). But by using `cast()` we effectively lose type checking:

```console
$ cat --number test.py
     1	#!/usr/bin/env python3
     2	
     3	from typing import TypedDict, cast
     4	
     5	class Struct(TypedDict):
     6	    foo: str
     7	    bar: int
     8	
     9	def test1() -> Struct:
    10	    return {'foo': 'foo', 'bar': 0}
    11	
    12	def test2() -> Struct:
    13	    return {'baz': None}
    14	
    15	def test3() -> Struct:
    16	    return cast(Struct, {'baz': None})
$ mypy test.py 
test.py:13: error: Missing keys ("foo", "bar") for TypedDict "Struct"  [typeddict-item]
test.py:13: error: Extra key "baz" for TypedDict "Struct"  [typeddict-unknown-key]
Found 2 errors in 1 file (checked 1 source file)
```

Does that mean that we need to avoid the use of `_asdict()` and `to_dict()` altogether? How annoying.

rdimaio replied:
Instead of using `cast`, this seems to work fine:
```
UsageDict(**session.execute(query).one()._asdict())
```

see test commit for reference: https://github.com/rdimaio/rucio/commit/cff7fa20131248acc2c4a06538c20b024d3c0b98

And pyright report: https://github.com/rdimaio/rucio/actions/runs/8233368875/job/22512678528

dchristidis replied:
I’m afraid that I failed to verify that it will raise an error in case of a mismatch:

```console
$ cat --number test.py 
     1	#!/usr/bin/env python3
     2	
     3	from typing import TypedDict
     4	
     5	from sqlalchemy import select
     6	from sqlalchemy.orm import Session
     7	
     8	from rucio.db.sqla.models import Identity
     9	from rucio.db.sqla.session import read_session
    10	
    11	
    12	# This is wrong.
    13	class IdentityDict(TypedDict):
    14	    foo: str
    15	    bar: int
    16	
    17	
    18	@read_session
    19	def list_identities(*, session: Session) -> list[IdentityDict]:
    20	    query = select(
    21	        Identity.identity
    22	    )
    23	    return [IdentityDict(**result._asdict()) for result in session.execute(query)]
    24	
    25	l = list_identities()
$ pyright test.py 
0 errors, 0 warnings, 0 informations
```

rdimaio replied:
This is odd, I found some sources online that seem to indicate that it should work:
- https://stackoverflow.com/questions/75254643/why-are-typeddict-types-without-field-and-with-notrequired-field-incompatible
- https://stackoverflow.com/questions/74495944/no-type-hint-with-return-values-for-typeddict-in-pycharm

And using extra keys in a TypedDict object construction (in the example, `**result._asdict()` are all extra keys to `IdentityDict`) should result in an error, according to PEP-589: https://stackoverflow.com/questions/71814658/python-typing-does-typeddict-allow-additional-extra-keys - the answer also mentions that pyright implements this.

I tried using a `dataclass` instead of a `TypedDict`, and I also don't get any errors raised from the incompatibility:

```
❯ cat --number test.py  
     1  #!/usr/bin/env python3
     2  from sqlalchemy import select
     3  from sqlalchemy.orm import Session
     4
     5  from rucio.db.sqla.models import Identity
     6  from rucio.db.sqla.session import read_session
     7  from dataclasses import dataclass
     8
     9  @dataclass
    10  class IdentityStruct:
    11      first: str
    12      second: int
    13
    14  @read_session
    15  def list_identities(*, session: Session) -> list[IdentityStruct]:
    16      query = select(
    17          Identity.identity
    18      )
    19      return [IdentityStruct(**result._asdict()) for result in session.execute(query)]
    20
    21  l = list_identities()                                                                                                                      


❯ python3 -m pyright test.py 
0 errors, 0 warnings, 0 informations 
```

rdimaio replied:
Some tests without the `sqla` components, just to check.

# 1. Incompatible TypedDict, constructor

```
#!/usr/bin/env python3
from typing import Any, TypedDict

class IncompatibleDict(TypedDict):
    foo: str
    bar: int

def print_dict(in_dict: dict[str, Any]) -> IncompatibleDict:
    d = IncompatibleDict(**in_dict)
    return(d)

test_dict = {
    "name": "test",
    "version": 3
}

l = print_dict(test_dict)
```

## Pyright doesn't complain
```
❯ python3 -m pyright test.py                                                                                                                      ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy complains that the dict type is unsupported for ** expansion
```
❯ python3 -m mypy test.py                                                                                                                         ─╯
test.py:9: error: Unsupported type "dict[str, Any]" for ** expansion in TypedDict  [typeddict-item]
Found 1 error in 1 file (checked 1 source file)
```

However, if I change the type annotation of `in_dict` from `dict[str, Any]` to `dict`, it behaves more as expected:
```
❯ python3 -m mypy test.py                                                                                                                         ─╯
test.py:9: error: Missing keys ("foo", "bar") for TypedDict "IncompatibleDict"  [typeddict-item]
Found 1 error in 1 file (checked 1 source file)
```

# 2. Incompatible TypedDict, cast
```
#!/usr/bin/env python3
from typing import Any, cast, TypedDict

class IncompatibleDict(TypedDict):
    foo: int
    bar: int

def print_dict(in_dict: dict[str, Any]) -> IncompatibleDict:
    d = cast(IncompatibleDict, in_dict)
    return(d)

test_dict = {
    "name": "test",
    "version": 3
}

l = print_dict(test_dict)
```

## Pyright doesn't complain
```
❯ python3 -m pyright test.py                                                                                                                      ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy doesn't complain
```
❯ python3 -m mypy test.py                                                                                                                         ─╯
Success: no issues found in 1 source file
```

# 3. Incompatible dataclass
```
#!/usr/bin/env python3
from typing import Any, cast
from dataclasses import dataclass

@dataclass
class IncompatibleClass:
    foo: int
    bar: int

def print_class(in_dict: dict[str, Any]) -> IncompatibleClass:
    d = cast(IncompatibleClass, in_dict)
    return(d)

test_dict = {
    "name": "test",
    "version": 3
}

l = print_class(test_dict)
```

## Pyright doesn't complain
```
❯ python3 -m pyright test.py                                                                                                                      ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy doesn't complain
```
❯ python3 -m mypy test.py                                                                                                                         ─╯
Success: no issues found in 1 source file
```

# Versions
```
❯ python3 -m pyright --version                                                                                                                    ─╯
pyright 1.1.353
```

```
❯ python3 -m mypy --version                                                                                                                       ─╯
mypy 1.9.0 (compiled: yes)
```

rdimaio replied:
# 4. Incompatible model, runtime Pydantic test, works as expected

```
#!/usr/bin/env python3
from typing import Any, cast
from pydantic import BaseModel, ConfigDict

class IncompatibleModel(BaseModel):
    foo: int
    bar: int
    model_config = ConfigDict(extra='forbid')

def print_class(in_dict: dict[str, Any]) -> IncompatibleModel:
    return(IncompatibleModel(**in_dict))

test_dict = {
    "name": "test",
    "version": 3
}

l = print_class(test_dict)
```

```
❯ python3 test.py                                                                                                                                 ─╯
Traceback (most recent call last):
  File "/home/rdm/tech/rucio/rucio/lib/rucio/core/test.py", line 18, in <module>
    l = print_class(test_dict)
  File "/home/rdm/tech/rucio/rucio/lib/rucio/core/test.py", line 11, in print_class
    return(IncompatibleModel(**in_dict))
  File "/home/rdm/.local/lib/python3.9/site-packages/pydantic/main.py", line 171, in __init__
    self.__pydantic_validator__.validate_python(data, self_instance=self)
pydantic_core._pydantic_core.ValidationError: 4 validation errors for IncompatibleModel
foo
  Field required [type=missing, input_value={'name': 'test', 'version': 3}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.6/v/missing
bar
  Field required [type=missing, input_value={'name': 'test', 'version': 3}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.6/v/missing
name
  Extra inputs are not permitted [type=extra_forbidden, input_value='test', input_type=str]
    For further information visit https://errors.pydantic.dev/2.6/v/extra_forbidden
version
  Extra inputs are not permitted [type=extra_forbidden, input_value=3, input_type=int]
    For further information visit https://errors.pydantic.dev/2.6/v/extra_forbidden
```

# 5. Compatible model, runtime Pydantic test, works as expected:
```
#!/usr/bin/env python3
from typing import Any, cast
from pydantic import BaseModel, ConfigDict

class CompatibleModel(BaseModel):
    name: str
    version: int
    model_config = ConfigDict(extra='forbid')

def print_class(in_dict: dict[str, Any]) -> CompatibleModel:
    return(CompatibleModel(**in_dict))

test_dict = {
    "name": "test",
    "version": 3
}

l = print_class(test_dict)
print(l)
```

```
❯ python3 test.py                                                                                                                                 ─╯
name='test' version=3
```

rdimaio replied:
Going back to static type checking - for test 4 (Incompatible model, runtime Pydantic test, works as expected), mypy and pyright do not detect issues by default:

```
❯ python3 -m mypy test.py                                                                                                                         ─╯
Success: no issues found in 1 source file
```

```
❯ python3 -m pyright test.py                                                                                                                      ─╯
0 errors, 0 warnings, 0 informations 
```

rdimaio replied:
# 6. Incompatible TypedDict, passing parameters directly, no extra parameters
```
#!/usr/bin/env python3
from typing import Any, TypedDict

class IncompatibleDict(TypedDict):
    foo: str
    bar: int

def print_dict() -> IncompatibleDict:
    test_dict = {
        "name": "test",
        "version": 3
    }
    d = IncompatibleDict(foo=test_dict['name'], bar=test_dict["version"])
    return(d)


l = print_dict()
```

## Pyright doesn't complain
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy complains
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:13: error: Incompatible types (expression has type "object", TypedDict item "foo" has type "str")  [typeddict-item]
lib/rucio/core/test.py:13: error: Incompatible types (expression has type "object", TypedDict item "bar" has type "int")  [typeddict-item]
Found 2 errors in 1 file (checked 1 source file)
```

# 7. Incompatible TypedDict, passing parameters directly, extra parameters
```
#!/usr/bin/env python3
from typing import Any, TypedDict

class IncompatibleDict(TypedDict):
    foo: str
    bar: int

def print_dict() -> IncompatibleDict:
    test_dict = {
        "name": "test",
        "version": 3
    }
    d = IncompatibleDict(foo=test_dict['name'], bar=test_dict["version"], extra="unexpected")
    return(d)


l = print_dict()
```

## Pyright complains
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
/home/rdm/tech/rucio/rucio/lib/rucio/core/test.py
  /home/rdm/tech/rucio/rucio/lib/rucio/core/test.py:13:9 - error: No overloads for "__init__" match the provided arguments
    Argument types: (Unknown, Unknown, Literal['unexpected']) (reportCallIssue)
1 error, 0 warnings, 0 informations 
```

## Mypy complains
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:13: error: Extra key "extra" for TypedDict "IncompatibleDict"  [typeddict-unknown-key]
lib/rucio/core/test.py:13: error: Incompatible types (expression has type "object", TypedDict item "foo" has type "str")  [typeddict-item]
lib/rucio/core/test.py:13: error: Incompatible types (expression has type "object", TypedDict item "bar" has type "int")  [typeddict-item]
Found 3 errors in 1 file (checked 1 source file)
```

rdimaio replied:
# 8. Incompatible Pydantic model, passing parameters directly, static type checking
```
#!/usr/bin/env python3
from pydantic import BaseModel, ConfigDict

class IncompatibleModel(BaseModel):
    foo: int
    bar: int

def print_class() -> IncompatibleModel:
    test_dict = {
        "name": "test",
        "version": 3
    }
    return(IncompatibleModel(foo=test_dict['name'], bar=test_dict['version']))

l = print_class()
```

Pydantic complains here, as we saw earlier. From the point of view of the static type checkers:

## Pyright doesn't complain
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy complains
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:13: error: Argument "foo" to "IncompatibleModel" has incompatible type "object"; expected "int"  [arg-type]
lib/rucio/core/test.py:13: error: Argument "bar" to "IncompatibleModel" has incompatible type "object"; expected "int"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

rdimaio replied:
# 9. Incompatible dataclass, passing parameters directly
```
#!/usr/bin/env python3
from typing import Any, cast
from dataclasses import dataclass, field

@dataclass
class IncompatibleClass:
    foo: int
    bar: int

def print_class() -> IncompatibleClass:
    test_dict = {
        "name": "test",
        "version": 3
    }
    d = IncompatibleClass(foo=test_dict['name'], bar=test_dict['version'])
    return(d)


l = print_class()
```

## Pyright doesn't complain
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
0 errors, 0 warnings, 0 informations 
```

## Mypy complains (ambiguously)
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:15: error: Argument "foo" to "IncompatibleClass" has incompatible type "object"; expected "int"  [arg-type]
lib/rucio/core/test.py:15: error: Argument "bar" to "IncompatibleClass" has incompatible type "object"; expected "int"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

Mypy complains even if I change `foo` to `str` - its problem seems to be that the type for `foo` and `bar` is not explicitly mentioned.

# 10. Incompatible dataclass, passing parameters directly
```
#!/usr/bin/env python3
from typing import Any, cast
from dataclasses import dataclass, field

@dataclass
class IncompatibleClass:
    foo: int
    bar: int

def print_class() -> IncompatibleClass:
    name = "test"
    version = 3
    d = IncompatibleClass(foo=name, bar=version)
    return(d)

l = print_class()
```

## Pyright complains
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
/home/rdm/tech/rucio/rucio/lib/rucio/core/test.py
  /home/rdm/tech/rucio/rucio/lib/rucio/core/test.py:13:31 - error: Argument of type "Literal['test']" cannot be assigned to parameter "foo" of type "int" in function "__init__"
    "Literal['test']" is incompatible with "int" (reportArgumentType)
1 error, 0 warnings, 0 informations 
```

## Mypy complains
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:13: error: Argument "foo" to "IncompatibleClass" has incompatible type "str"; expected "int"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

# 11. Incompatible TypedDict, passing parameters directly, adding explicitly types for the input arguments
```
#!/usr/bin/env python3
from typing import Any, TypedDict

class IncompatibleDict(TypedDict):
    foo: int
    bar: int

def print_dict() -> IncompatibleDict:
    name: str = "test"
    version: int = 3
    d = IncompatibleDict(foo=name, bar=version)
    return(d)


l = print_dict()
```

## Pyright complains
```
❯ python3 -m pyright lib/rucio/core/test.py                                                                                                       ─╯
/home/rdm/tech/rucio/rucio/lib/rucio/core/test.py
  /home/rdm/tech/rucio/rucio/lib/rucio/core/test.py:11:30 - error: Argument of type "Literal['test']" cannot be assigned to parameter "foo" of type "int" in function "__init__"
    "Literal['test']" is incompatible with "int" (reportArgumentType)
1 error, 0 warnings, 0 informations 
```

## Mypy complains
```
❯ python3 -m mypy lib/rucio/core/test.py                                                                                                          ─╯
lib/rucio/core/test.py:11: error: Incompatible types (expression has type "str", TypedDict item "foo" has type "int")  [typeddict-item]
Found 1 error in 1 file (checked 1 source file)
```

rdimaio replied:
# Conclusion
- Static type checking: mypy seems to be slightly stricter compared to pyright - I think it would be worth considering we move to mypy.
- Runtime type checking: I think this should be necessary to catch cases where static type checking would not be enough. I opened an issue to discuss this: https://github.com/rucio/rucio/issues/6544

In the long term, I think the best approach would be to use mypy+pydantic and use pydantic models for struct typing. For now, it seems to be fine to use `TypedDict`, passing arguments directly, instead of passing them as kwargs, and adding explicit type annotations for the input arguments, so that the static type checkers are able to notice any issues.

rdimaio replied:
Relevant thread https://github.com/rucio/rucio/discussions/6615

dchristidis commented on lib/rucio/core/account.py:
```diff
@@ -373,13 +374,13 @@ def get_usage(rse_id, account, *, session: "Session"):
-        return session.execute(query).one()._asdict()
+        return cast(UsageDict, session.execute(query).one()._asdict())
     except exc.NoResultFound:
-        return {'bytes': 0, 'files': 0, 'updated_at': None}
+        return UsageDict({'bytes': 0, 'files': 0, 'updated_at': None})
```
Similarly here.

rdimaio replied:
Replied here: https://github.com/rucio/rucio/pull/6497#discussion_r1518058114

dchristidis commented on lib/rucio/core/account_counter.py:
```diff
@@ -12,12 +12,14 @@
 # limitations under the License.
+
 import datetime
-from typing import TYPE_CHECKING
+from typing import cast, TYPE_CHECKING
```
Why do you hate sorting? Why do you hate order? :laughing:

rdimaio replied:
What's the issue here?

dchristidis replied:
```console
$ isort --diff sort.py 
--- /var/home/dchristidis/sort.py:before	2024-03-08 20:42:16.143365
+++ /var/home/dchristidis/sort.py:after	2024-03-08 20:42:28.548160
@@ -1 +1 @@
-from typing import cast, TYPE_CHECKING
+from typing import TYPE_CHECKING, cast
```

dchristidis commented on lib/rucio/core/account_limit.py:
```diff
@@ -268,7 +273,7 @@ def delete_global_account_limit(account, rse_expression, *, session: "Session"):
 
 
 @transactional_session
-def get_local_account_usage(account, rse_id=None, *, session: "Session"):
+def get_local_account_usage(account: InternalAccount, rse_id: Optional[UUID] = None, *, session: "Session") -> list[RSELocalAccountUsageDict]:
```
I still think that `UUID` is wrong and I would like to get to the bottom of this.

dchristidis commented on lib/rucio/core/authentication.py:
```diff
@@ -30,17 +30,17 @@
 from rucio.common.utils import chunks, generate_uuid, date_to_str
 from rucio.core.account import account_exists
-from rucio.core.oidc import validate_jwt
+from rucio.core.oidc import validate_jwt, token_dictionary
```
Forgive me, it should be the other way around: the `token_dictionary()` in `rucio.core.authentication` should be kept and the one in `rucio.core.oidc` replaced.

rdimaio replied:
This causes a circular import:

```
ImportError: cannot import name 'validate_jwt' from partially initialized module 'rucio.core.oidc' (most likely due to a circular import) (/opt/venv/lib64/python3.9/site-packages/rucio/core/oidc.py)
```

execution for reference: https://github.com/rucio/rucio/actions/runs/8207133900/job/22447782460?pr=6497

rdimaio commented:
Sorry for leaving this unattended until now.

I am breaking down this PR into smaller PRs to help move things forward. Each PR will be for a different aspect of this PR:
- One PR for the `ssh_sign` changes: https://github.com/rucio/rucio/pull/6647
- One for the `get_auth_token_ssh` changes (in this one, we can explore this proposal further: https://github.com/rucio/rucio/pull/6497#discussion_r1516152819): https://github.com/rucio/rucio/pull/6649
- For the type hints (split into one per file):
  - authentication.py: https://github.com/rucio/rucio/pull/6742

Some PRs linked above have already been merged. I am closing this PR.
